numpy/core/src/umath/matmul.c.src

```diff
@@ -494,9 +499,56 @@ NPY_NO_EXPORT void
                  * non-blasable (or non-ccontiguous output)
                  * we could still use BLAS, see gh-12365.
                  */
-                @TYPE@_matmul_inner_noblas(ip1, is1_m, is1_n, 
-                                           ip2, is2_n, is2_p,
-                                           op, os_m, os_p, dm, dn, dp);
+                if(tmp_ip1 == NULL) {
+                    tmp_ip1 = (@typ@ *)malloc(sizeof(@typ@) * dn * dm);
+                }
+                if(tmp_ip2 == NULL) {
+                    tmp_ip2 = (@typ@ *)malloc(sizeof(@typ@) * dn * dp);
+                }
+                if(tmp_op == NULL) {
+                    tmp_op = (@typ@ *)malloc(sizeof(@typ@) * dm * dp);
+                }
+
+                /* Not enough memory */
+                if(
+                    tmp_ip1 == NULL || tmp_ip2 == NULL || tmp_op == NULL
+                ) {
+                    @TYPE@_matmul_inner_noblas(ip1, is1_m, is1_n, 
+                                               ip2, is2_n, is2_p,
+                                               op, os_m, os_p, dm, dn, dp);
+                }
+                else {
+                    for (n = 0; n < dn; n++) {
+                        for (m = 0; m < dm; m++) {
+                            tmp_ip1[m*dn + n] =
+                            *(@typ@ *)(((char *)ip1) + m*is1_m + n*is1_n);
+                        }
+                        for (p = 0; p < dp; p++) {
+                            tmp_ip2[n*dp + p] =
+                            *(@typ@ *)(((char *)ip2) + p*is2_p + n*is2_n);
+                        }
+                    }
+                    for (m = 0; m < dm; m++) {
+                        for (p = 0; p < dp; p++) {
+                            tmp_op[m*dp + p] =
+                            *(@typ@ *)(((char *)op) + m*os_m + p*os_p);
+                        }
+                    }
```

Comment:
Added a new function `@TYPE@_matrix_copy`. Pretty generic for 2D matrices. However, I tried to include `malloc` and as far as I can judge, it does not make too much sense, because two copies are in one direction (original arrays to temp arrays) and one copy is in the opposite direction (temp array to original array). When trying to malloc within the copy function, it messes things up.